src/agreement/mod.rs

```diff
@@ -210,39 +208,43 @@ pub struct Agreement<NodeUid> {
     coin_schedule: CoinSchedule,
 }
 
+pub type AgreementStep<N> = Step<N, bool>;
+
 impl<NodeUid: Clone + Debug + Ord> DistAlgorithm for Agreement<NodeUid> {
     type NodeUid = NodeUid;
     type Input = bool;
     type Output = bool;
     type Message = AgreementMessage;
     type Error = Error;
 
-    fn input(&mut self, input: Self::Input) -> AgreementResult<FaultLog<NodeUid>> {
-        self.set_input(input)
+    fn input(&mut self, input: Self::Input) -> AgreementResult<AgreementStep<NodeUid>> {
+        let fault_log = self.set_input(input)?;
+        self.step(fault_log)
     }
 
     /// Receive input from a remote node.
     fn handle_message(
         &mut self,
         sender_id: &Self::NodeUid,
         message: Self::Message,
-    ) -> AgreementResult<FaultLog<NodeUid>> {
-        if self.terminated || message.epoch < self.epoch {
-            // Message is obsolete; we're already in a later epoch or terminated.
-            return Ok(FaultLog::new());
-        }
-        if message.epoch > self.epoch {
+    ) -> AgreementResult<AgreementStep<NodeUid>> {
+        let fault_log = if self.terminated || message.epoch < self.epoch {
+            // Message is obsolete: We are already in a later epoch or terminated.
+            FaultLog::default()
```

Comment:
You could use `FaultLog::new()` here, which just calls `default()`. Same goes for the `else if` branch.